proposed/dynamic-component-loading.md

```diff
@@ -0,0 +1,110 @@
+    * This provides full isolation to the component. Every dependency the component carries with it will be used from the component and loaded in isolation from the rest of the app. Basically avoids any potential collisions.
+    * On the downside, this doesn't provide implicit sharing. Typically only framework assemblies would be shared in this scenario. Component would have to explicitly choose which assemblies to share, by not carrying them with it (this can be setup in project file by using `CopyLocal=false` for assembly references, similar option exists for project and NuGet references as well). This means that types used for communication between the app and the component would have to be explicitly shared by the component (via the exclusion of the assembly in the component). This is not done by the SDK by default, so it's easy to get this wrong and even with improved diagnostics will be relatively hard to debug.
+* **Always load into default** - in this case all loads are done to the parent (Default) load context. In the extreme case a new load context is not really needed. Alternative could be that the load is attempted to the default load context (first by name, then by resolved file path). If that fails (should really only happen in case of a collision), the dependency would be loaded into the new load context in isolation.
+    * Inherently shares as much as possible - avoids problems of sharing types used for communication.
+    * Downside is that loading a component will "pollute" the default load context with assemblies from the component.
+    * Loading multiple components with similar dependencies can easily lead to unpredictable results as what gets loaded where would depend on ordering.
+    * Auto-upgrades - if the app uses a higher version of a given dependency, component which uses a lower version of the dependency will get the app's version - this can lead to incompatibilities.
+* **Prefer default** - in this case a new load context is created. When it tries to resolve a dependency, it will first try to resolve it against the parent (Default) load context. If the parent can satisfy the dependency, it will be used. Otherwise if the dependency can't be resolved by the parent, the dependency will be resolved to a file path and loaded in isolation into the new load context.
+    * Inherently shares all assemblies from the parent (Default) - avoids problems with sharing types used for communication.
+    * There's no pollution of the parent context - no new files will be loaded into the parent context.
+    * Auto-upgrade - components will auto-upgrade to the version of a dependency they share with the app (downgrade will not occur, in that case the dependency would be loaded in isolation). This can lead to incompatibilities.
+
+Open questions:
+* Which is the default behavior?
```

Comment:
Like class loaders in Java? Perhaps than assemblies loading can have an API as well so it may decide whether to give a chance to parent to load assembly first. The same way as class loaders works.

Reply:
@SergeyKanzhelev I have only very limited knowledge of Java in general. So far what I've read about class loaders they are very similar to .NET Core's `AssemblyLoadContext`. This class already provides ways for a custom implementation to defer to the parent first if the custom code wants to do that.  

This proposal heavily relies on `AssemblyLoadContext` as the way to achieve isolation. I guess one way of thinking about this proposal is that we're trying to add an easier-to-use class loader to dynamically load components with their dependencies. But again, I know very little about the Java world, so maybe I'm completely wrong.

Reply:
I need to read about `AssemblyLoadContext` more.

Yes, I was mostly referring to the fact that class loaders can be custom coded to decide on loading ordering themselves. And also there are few well-known class loaders which user is aware of and can decide to load some assemblies there so they will share statics.

If simplified version if enough for most scenarios - it should be fine.

Reply:
There's a description of [`AssemblyLoadContext`](https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/assemblyloadcontext.md). It's not entirely up-to-date, but the basics are there. Also added this link to the doc. Let me know if there are unclear parts, I'll go fix the other doc as well.

Reply:
I know my comment won't be of any help in terms of how you guys would proceed but I just wanted to say, with 20 or so many years of software development background, the `AssemblyLoadContext` is very hard to understand. Either due to the API design or due to the lack of documentation (samples rather than API description), I don't know. Just wanted to share how I feel about it.

Reply:
@CoskunSunali Thanks.  We are hearing the same thing from lots of customers.